### PR TITLE
Improve Try error messages

### PR DESCRIPTION
**Pull request by huntiep**
This is a WIP. Fixing `ui/suggestions/try-operator-on-main.rs` requires implementing the `Return type does not implement Try` situation of [RFC 1859](https://github.com/rust-lang/rfcs/blob/master/text/1859-try-trait.md#error-messages). This is part of #35946.

How would I differentiate between the first two situations of the RFC?

### CODE REVIEW

**Comment by pnkfelix** [comment]
@huntiep just to be clear, when you asked "How would I differentiate between the first two situations of the RFC?", are you asking how to differentiate:

 * Source type does not implement Try, vs
 * Return type does not implement Try

?

**Comment by huntiep** [comment]
@pnkfelix Yes, sorry for not being clear.

**Comment by carols10cents** [comment]
ping @pnkfelix, do you have an answer for @huntiep's question now that they've clarified?

**Comment by nikomatsakis** [comment]
ah, er, didn't meant to steal from @pnkfelix :) either way is fine. @pnkfelix, I missed that you had left some comments.

File: src/test/ui/suggestions/try-unimplemented.stderr [hunk 1]
```diff
@@ -0,0 +1,26 @@
+error[E0277]: the trait bound `u32: std::ops::Try` is not satisfied
+  --> $DIR/try-unimplemented.rs:17:5
+   |
+17 |     3u32?
+   |     -----
+   |     |
+   |     `?` cannot be applied to a value of type `u32`
+   |     in this macro invocation
```
Comment (nikomatsakis) [inline src/test/ui/suggestions/try-unimplemented.stderr]: Hmm, we have to get rid of this "in this macro invocation"....

**Comment by nikomatsakis** [comment]
@huntiep hmm, that's a good question! I think I overlooked that point in my mentoring instructions. Really, the easiest way to track this would be if we can further use the span. In particular, if we could give a distinct desugaring code to the "return" or "break" generated by the `?` desugaring, it seems like would work.

File: src/librustc_errors/emitter.rs [hunk 2]
```diff
@@ -724,7 +724,9 @@ impl EmitterWriter {
 
             // First, find all the spans in <*macros> and point instead at their use site
             for sp in span.primary_spans() {
-                if *sp == DUMMY_SP {
+                if *sp == DUMMY_SP ||
+                    sp.is_compiler_desugaring(CompilerDesugaringKind::QuestionMark)
```
Comment (nikomatsakis) [inline src/librustc_errors/emitter.rs]: Cool =)

File: src/test/ui/suggestions/try-unimplemented.stderr [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+   |     ^^^^^ `?` cannot be applied to a value of type `u32`
+   |
+   = help: the trait `std::ops::Try` is not implemented for `u32`
+   = note: required by `std::ops::Try::into_result`
+
+error[E0277]: the trait bound `(): std::ops::Try` is not satisfied
+  --> $DIR/try-unimplemented.rs:21:5
+   |
+21 |     Ok(3u32)?;
+   |     ^^^^^^^^^ cannot use the `?` operator in a function that returns `()`
```
Comment (nikomatsakis) [inline src/test/ui/suggestions/try-unimplemented.stderr]: is this...aspirational?

Comment (nikomatsakis) [reply src/test/ui/suggestions/try-unimplemented.stderr]: i.e., I don't see code that would emit this "cannot use" error

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@huntiep I took a shot at elaborating what I had in mind for distinguishing the two cases here: https://github.com/rust-lang/rust/issues/35946#issuecomment-325655073

**Comment by arielb1** [comment]
Superseded by #44191, which improves `on_unimplemented` while setting the stage for better error reporting for other builtin traits.

File: src/test/ui/suggestions/try-unimplemented.stderr (hunk 3 above)
Comment (huntiep) [reply src/test/ui/suggestions/try-unimplemented.stderr]: It was intended to test `Return type does not implement Try`. Looking at `try-operator-on-main`, it has a pretty similiar error message, so maybe it was decided to use that error message instead.

**Comment by huntiep** [comment]
@arielb1 Should I close this then?

**Comment by arielb1** [comment]
@huntiep

Sure.

**Pull request closed without merging**